### DISCUSSION TITLE
[Android] NativeControlHost Airspace

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#13953] [Q&A]
Hi, I'm trying to implement `CameraView` for Android. To do this I sublated `NativeControlHost` and implemented the binding with the native view. This works quite well except for `Airspace`, the Avalonia controls are always below the native control. Do you have any suggestions on how to solve it or suggest different ways?

Avalonia Version **11.0.6**

**Comment by timunie** [comment]
I think this is by design. Afaik only option you have is to add another AvaloniaControlHost inside NativeControlHost which will then render Avalonia over Native over Avalonia 🙃

**Reply by workgroupengineering** [reply]
Hi, @timunie thank you for answering.

I will try your suggestion, but the main problem is with the flyouts. See following image.

![immagine](https://github.com/AvaloniaUI/Avalonia/assets/12531229/befc7ef6-e6ee-45c6-8f05-3855d6802993)

**Reply by timunie** [reply]
In WPF I ended up hiding the native control for the time a Popup was open. Not sure if that could be a workaround here?

**Reply by workgroupengineering** [reply]
From what I understand Avalonia renders all UI on layer 0 of the native surface. When adding `NativeControlHost` an additional layer is created on top of the one where Avalonia renders, this causes the issue.

A possible solution only for controls like Flyout could be that when an overlay is created a new native layer is added on top of the existing ones.

Another solution could be to render the native controls in an invisible layer and create a snapshot to send to Avalonia.

What do you think, which one is the most feasible in your opinion?

**Comment by workgroupengineering** [comment]
Hi, @kekekeks 
Sorry to ping you, but I want to avoid investing time on a solution to this problem that isn't shared by the development team.

- in android we could create a [virtual desktop](https://developer.android.com/reference/android/hardware/display/VirtualDisplay) with the same characteristics as the one on which the Avaonia app is run, on which you will render the native controls and then create a snapshort to be rederig to avalonia.

- I noticed that today ITopLevelImpl can have one or more Surfaces. we might think about implementing mark interfaces that identify the following types of surfaces e.g.:
   ```csharp

      // Identifies the main surfaces where the composer renders the UI. (Layer 0)
      public IAvaloniaMainSurfaceGroup{}

      // Surface where the native Controls are render (Layer 1)
      public IAvalionaNativeControlsHostSurfaceGroup{}

      // Surface where all Avalonia overlays (Flyout, Popup, AdornerLayer, ...) are rendered (Layer 2)
      public IAvalionaOverlaySurfaceGroup{}

   ```

**Comment by workgroupengineering** [comment]
I was able to place the native controls on a layer below the one where Avalonia renders. How can I tell the Avalonia rendering engine not to draw anything in areas where NativeControlHost is present?

![immagine](https://github.com/AvaloniaUI/Avalonia/assets/12531229/1ba6a7cf-1db7-48a9-bae1-a9bb4e106be9)

**Reply by workgroupengineering** [reply]
> I was able to place the native controls on a layer below the one where Avalonia renders. How can I tell the Avalonia rendering engine not to draw anything in areas where NativeControlHost is present?
> 
> ![immagine](https://private-user-images.githubusercontent.com/12531229/293054638-1ba6a7cf-1db7-48a9-bae1-a9bb4e106be9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3NTAwNjYsIm5iZiI6MTcwMzc0OTc2NiwicGF0aCI6Ii8xMjUzMTIyOS8yOTMwNTQ2MzgtMWJhNmE3Y2YtMWRiNy00OGE5LWJhZTEtYTliYjRlMTA2YmU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDA3NDkyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiZjg1MzQ1YTYxODUyMmZlNWJmNjEwZTI3NDRlNTY2MDRhODY2MTRiODg4YTIzYzYyYjkxZjBkZjgyNzFkYmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3kSyeGjQWL1dlEPXlTO9XORWHYNnMU6SEnNHvUrdzTI)

Hi, @maxkatz6 
Do you have any suggestions on how to try to resolve this? I tried with the Clip but so far I haven't achieved the desired result.

**Reply by workgroupengineering** [reply]
I finally achieve to clipping TopLevel but I have a issue. Is there a way to avoid clipping `OverlayPopupHost` when clipping TopLevel?

**Comment by msneijders** [comment]
@workgroupengineering did you find a workable solution for this? I am also facing a challenge to get a NativeControlHost (I also want a live camera feed, but also a browser) to be more "part of" the avalonia app instead of just floating on top of it.

Just as you I was thinking about multiple layers/surfaces (main content on the bottom, a surface for offcanvas menu and a bottomsheet on top, and the nativecontrolhost in the middle).

**Reply by workgroupengineering** [reply]
Unfortunately not, the project it was for has become low priority.  Without the support of the core team it is difficult to find a solution.

**Reply by msneijders** [reply]
:`-(

**Reply by timunie** [reply]
if you have a support contract, reach out to the team. However I feel like there is no simple solution for that usecase (also not in other UI Frameworks)
